### DISCUSSION TITLE
Change token claims dynamically after signin

### DISCUSSION THREAD

**Discussion by techpet** [nhost/nhost#1589] [Q&A]
In our architecture, the nhost user account is connected with a profile which could be a member of multiple workspaces (having different roles per workspace). In our app, a user could select a working workspace from the ones that he is a member in and then have access to many different entities connected with the specific workspace based on his role on this. So, the access to related entities is based on the workspace id and the user (profile) role in this workspace. 

How could the `x-hasura-selected-worspace-id` be specified dynamically as a session variable based on the user workspace selection so that each time that the user changes the selected workspace, the `x-hasura-selected-workspace-id` parameter is updated accordingly?

What is a recommended way of achieving this functionality?

In theory, a new, custom endpoint could be implemented that would verify the access token, read the value of the selected workspace id as a param and then create a new token with a custom claim: `x-hasura-selected-worspace-id` set to the param id value. But how would the refresh token mechanism and the  `nhost.auth.refreshToken()` work with this afterwards?

Would we have to override the default nhost auth sdk tokenHandler functionality also?

**Comment by elitan** [comment]
Hmm, I'm actually not sure how to handle this. If you, or anyone else, have some ideas, please write here. I'll keep thinking about how this could be handled.

Are there any reference service that handles this well? How do they do it?

**Comment by techpet** [comment]
OK, the way that we've solved this is by adding an additional `selectedWorkspace` field in the profile table that holds the id of the selected workspace and this field is configured in nhost custom claims .env parameter. 

Whenever the user changes his active workspace client side, the selected workspace id is sent in a profile mutation to Hasura that alters the `selectedWorkspace`  field value. Upon success,  `nhost.auth.refreshToken()` is called to obtain a new JWT with the updated selected workspace id value and afterwards this is used in follow up requests to provide access only to related entities filtered in Hasura permissions using the custom claim.

Does it sound like a valid approach?
